File: backend/app/features/vpn/schemas/topology.py
from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict


class VpnServerRead(BaseModel):
    interface: str = "wg0"
    endpoint: str
    gateway_ip: str
    dns_ip: str
    cidr: str
    server_public_key: str
    allowed_ips: List[str]
    mtu: Optional[int] = None


class VpnPeerNodeRead(BaseModel):
    peer_id: int
    vpn_device_id: str
    device_id: Optional[int] = None
    hostname: Optional[str] = None
    client_ip: str
    public_key: str
    handshake_status: Literal["connected", "idle", "never", "unknown"] = "unknown"
    latest_handshake_at: Optional[datetime] = None
    endpoint: Optional[str] = None
    rx_bytes: Optional[int] = None
    tx_bytes: Optional[int] = None
    on_wireguard: bool = False

    model_config = ConfigDict(from_attributes=True)


class VpnTopologyRead(BaseModel):
    server: VpnServerRead
    peers: List[VpnPeerNodeRead]

File: frontend/src/features/vpn/types/topology.ts
export type HandshakeStatus = 'connected' | 'idle' | 'never' | 'unknown';

export interface VpnServer {
  interface: string;
  endpoint: string;
  gateway_ip: string;
  dns_ip: string;
  cidr: string;
  server_public_key: string;
  allowed_ips: string[];
  mtu: number | null;
}

export interface VpnPeerNode {
  peer_id: number;
  vpn_device_id: string;
  device_id: number | null;
  hostname: string | null;
  client_ip: string;
  public_key: string;
  handshake_status: HandshakeStatus;
  latest_handshake_at: string | null;
  endpoint: string | null;
  rx_bytes: number | null;
  tx_bytes: number | null;
  on_wireguard: boolean;
}

export interface VpnTopology {
  server: VpnServer;
  peers: VpnPeerNode[];
}
